### PR TITLE
fix(fields): handle properly default value for type `Callable`

### LINKED ISSUE #1596: Callables are attached as bound methods when used as default args
# Bug

Methods used as default args to Callable are attached as bound methods of the BaseModel class.  Things work fine if you pass the method as an argument to the constructor.  One workaround (if you must have a default arg) is to pass the method as a `functools.partial` object.

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.5.1
            pydantic compiled: True
                 install path: /usr/local/lib/python3.7/site-packages/pydantic
               python version: 3.7.7 (default, May 20 2020, 21:10:21)  [GCC 8.3.0]
                     platform: Linux-4.19.76-linuxkit-x86_64-with-debian-10.4
     optional deps. installed: ['typing-extensions']
```

### Test

```py
from pydantic import BaseModel
import typing as t
from functools import partial

def bar(*args):
    print(f"bar args: {args}")

class Foo(BaseModel):
    callback: t.Callable[[int], int]
    method: t.Callable[..., t.Any] = bar

m = Foo(callback=lambda x: x)
print(m.callback(42))
print(m.method)
m.method(42)
print(f"\n\n{'#' * 42}\n\n")

class Foo2(BaseModel):
    callback: t.Callable[[int], int]
    method: t.Callable[..., t.Any]  # no default

n = Foo2(callback=lambda x: x, method=bar)  # This works fine
print(n.callback(42))
print(n.method)
n.method(42)
print(f"\n\n{'#' * 42}\n\n")

class Foo3(BaseModel):
    callback: t.Callable[[int], int]
    method: t.Callable[..., t.Any] = partial(bar)  # workaround

o = Foo3(callback=lambda x: x)
print(o.callback(42))
print(o.method)
o.method(42)
```

### Output

```shell
# python foo.py 
42
<bound method bar of Foo(callback=<function <lambda> at 0x7f596462b9e0>)>
bar args: (Foo(callback=<function <lambda> at 0x7f596462b9e0>), 42)


##########################################


42
<function bar at 0x7f5964621dd0>
bar args: (42,)


##########################################


42
functools.partial(<function bar at 0x7f5964621dd0>)
bar args: (42,)
```

### LINKED ISSUE #2086: TypeError during generation schema for model field (type Callable) with provided default value.
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.7.1
            pydantic compiled: True
                 install path: /home/***/.local/share/virtualenvs/lerdem-oY9IIzMI/lib/python3.6/site-packages/pydantic
               python version: 3.6.9 (default, Oct  8 2020, 12:12:24)  [GCC 8.4.0]
                     platform: Linux-5.0.0-32-generic-x86_64-with-LinuxMint-19.3-tricia
     optional deps. installed: []
```

```py
>>> from pydantic import BaseModel, Field
>>> from typing import Callable
>>> class Foo(BaseModel):
...     # without default value, getting user warning
...     callback: Callable[[], dict]
... 
>>> Foo.schema()
__main__:1: UserWarning: Callable callback was excluded from schema since JSON schema has no equivalent type.
{'title': 'Foo', 'type': 'object', 'properties': {}}

>>> class FooDefault(BaseModel):
...     callback: Callable[[], dict] = Field(default=lambda: {'some': 'data'})
>>> FooDefault.schema()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 637, in pydantic.main.BaseModel.schema
  File "pydantic/schema.py", line 153, in pydantic.schema.model_schema
  File "pydantic/schema.py", line 523, in pydantic.schema.model_process_schema
  File "pydantic/schema.py", line 564, in pydantic.schema.model_type_schema
  File "pydantic/schema.py", line 215, in pydantic.schema.field_schema
  File "pydantic/schema.py", line 183, in pydantic.schema.get_field_info_schema
  File "pydantic/schema.py", line 846, in pydantic.schema.encode_default
  File "pydantic/json.py", line 65, in pydantic.json.pydantic_encoder
TypeError: Object of type 'function' is not JSON serializable
```

One possible solution according for docs is to use custom [`Confing.json_encoders`](https://pydantic-docs.helpmanual.io/usage/exporting_models/#json_encoders). So i have tried:
```py
>>> def hack():
...     pass
>>> class FooDefault(BaseModel):
...     callback: Callable[[], dict] = Field(default=lambda: {'some': 'data'})
...     class Config:
...         json_encoders = {
...             # i don't know how to make it right way
...             type(hack): str,
...             Callable: str
...         }
>>> FooDefault.schema()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 637, in pydantic.main.BaseModel.schema
  File "pydantic/schema.py", line 153, in pydantic.schema.model_schema
  File "pydantic/schema.py", line 523, in pydantic.schema.model_process_schema
  File "pydantic/schema.py", line 564, in pydantic.schema.model_type_schema
  File "pydantic/schema.py", line 215, in pydantic.schema.field_schema
  File "pydantic/schema.py", line 183, in pydantic.schema.get_field_info_schema
  File "pydantic/schema.py", line 846, in pydantic.schema.encode_default
  File "pydantic/json.py", line 65, in pydantic.json.pydantic_encoder
TypeError: Object of type 'function' is not JSON serializable
```
I think, that preferable behaviour is to show warnings in situation like that. And that should work without any custom `json_encoders`.

### PR DESCRIPTION
**Pull request by PrettyWood**
## Change Summary
```python
from typing import Callable

from pydantic import BaseModel


def pika_default() -> str:
    return 'Pika is the best'

class PikaModel(BaseModel):
    pika: Callable[[], str] = pika_default


print(PikaModel.__fields__)
```

#### BEFORE
```
{}
```

#### AFTER
```
{'pika': ModelField(name='pika', type=Callable[[], str], required=False, default=<function pika_default at 0x10a6541f0>)}
```

## Related issue number
closes #1596
closes #2086

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

File: tests/test_main.py [hunk 1]
```diff
@@ -1374,3 +1376,47 @@ class M(BaseModel):
         a: int
 
     get_type_hints(M.__config__)
+
+
+def foo(arg1, arg2):
```
Comment (samuelcolvin) [inline tests/test_main.py:1381]: could we move this out of `test_main.py` e.g. into `test_dataclasses.py` or `test_edge_cases.py`?

File: tests/test_main.py [hunk 2]
```diff
@@ -1374,3 +1376,47 @@ class M(BaseModel):
...
+    return arg1, arg2
+
+
+@pydantic_dataclass
```
Comment (samuelcolvin) [inline tests/test_main.py:1385]: where possible please define models and pydantic dataclasses inside tests. Most of the logic we're testing here occurs then the type is created.

File: pydantic/main.py [hunk 3]
```diff
@@ -198,7 +198,8 @@ def validate_custom_root_type(fields: Dict[str, ModelField]) -> None:
         raise ValueError('__root__ cannot be mixed with other fields')
 
 
-UNTOUCHED_TYPES = FunctionType, property, type, classmethod, staticmethod
+FIELD_DEFAULT_VALUE_UNTOUCHED_TYPES = property, type, classmethod, staticmethod
```
[the remark this answers is not in the document]
Comment (PrettyWood) [reply pydantic/main.py]: I added a comment and renamed it to `ANNOTATED_FIELD_UNTOUCHED_TYPES`. Still a bit long but I couldn't find an explicit name shorter than that.

File: tests/test_main.py (hunk 1 above)
Comment (PrettyWood) [reply tests/test_main.py]: Moved to `test_dataclasses` with model creations inside the test

File: tests/test_dataclasses.py [hunk 4]
```diff
@@ -801,6 +801,46 @@ class Config:
...
+    @dataclasses.dataclass
+    class HasCallablesStdlibDC:
+        non_default_callable: Callable
+        default_callable: Callable = lambda x: foo(x, 'default')
+        default_callable_factory: Callable = dataclasses.field(default_factory=lambda: lambda x: foo(x, 'factory'))
+
+    def non_default_callable(x):
+        return foo(x, 'nondefault')
+
+    for cls in (HasCallablesModel, HasCallablesDC):
```
Comment (samuelcolvin) [inline tests/test_dataclasses.py:830]: it's really confusing if something fails to work out which object failed.

Better just to split the test and duplicate code, or if there's a lot of duplication, use fixtures.

File: tests/test_dataclasses.py [hunk 5]
```diff
@@ -801,6 +801,46 @@ class Config:
...
+        return arg1, arg2
+
+    @pydantic.dataclasses.dataclass
+    class HasCallablesDC:
+        non_default_callable: Callable
+        default_callable: Callable = lambda x: foo(x, 'default')
+        default_callable_factory: Callable = dataclasses.field(default=lambda x: foo(x, 'factory'))
+
+    class HasCallablesModel(BaseModel):
+        non_default_callable: Callable
```
Comment (samuelcolvin) [inline tests/test_dataclasses.py:817]: what about typed callables like `Callable[[int, float, str], bool]`?

Surely worth adding tests for them too?

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM

File: tests/test_dataclasses.py (hunk 4 above)
Comment (PrettyWood) [reply tests/test_dataclasses.py]: I copied the test from the issue but I agree it was a bit confusing. I rewrote the test. Hope it's easier to read.
Also added with an annotated callable like `Callable[[int, float, str], bool]`.
Ready for another review :)

**Comment by PrettyWood** [comment]
Added a small commit for schema generation in the batch since it's related (callable fields with default value) even though it's not the same root cause 😉

**Pull request merged**